src/Xamarin.Android.Build.Tasks/Utilities/LlvmIrGenerator/LlvmIrStringManager.cs

```diff
@@ -1,43 +1,64 @@
 using System;
 using System.Collections.Generic;
 
+using Microsoft.Build.Utilities;
+
 namespace Xamarin.Android.Tasks.LLVMIR;
 
 partial class LlvmIrModule
 {
 	protected class LlvmIrStringManager
 	{
-		Dictionary<string, LlvmIrStringVariable> stringSymbolCache = new Dictionary<string, LlvmIrStringVariable> (StringComparer.Ordinal);
+		readonly string defaultGroupName = "str";
+
+		Dictionary<StringHolder, LlvmIrStringVariable> stringSymbolCache = new Dictionary<StringHolder, LlvmIrStringVariable> ();
 		Dictionary<string, LlvmIrStringGroup> stringGroupCache = new Dictionary<string, LlvmIrStringGroup> (StringComparer.Ordinal);
 		List<LlvmIrStringGroup> stringGroups = new List<LlvmIrStringGroup> ();
 
 		LlvmIrStringGroup defaultGroup;
+		TaskLoggingHelper log;
 
 		public List<LlvmIrStringGroup> StringGroups => stringGroups;
 
-		public LlvmIrStringManager ()
+		public LlvmIrStringManager (TaskLoggingHelper log, string? defaultStringGroup = null)
 		{
+			this.log = log;
+			if (!String.IsNullOrEmpty (defaultStringGroup)) {
+				defaultGroupName = defaultStringGroup;
+			}
+
 			defaultGroup = new LlvmIrStringGroup ();
 			stringGroupCache.Add (String.Empty, defaultGroup);
 			stringGroups.Add (defaultGroup);
 		}
 
-		public LlvmIrStringVariable Add (string value, string? groupName = null, string? groupComment = null, string? symbolSuffix = null)
+		public LlvmIrStringVariable Add (LlvmIrStringVariable variable, string? groupName = null, string? groupComment = null, string? symbolSuffix = null)
+		{
+			// Let it throw if Value isn't a StringHolder, it must be.
+			return Add((StringHolder)variable.Value, groupName, groupComment, symbolSuffix);
+		}
+
+		public LlvmIrStringVariable Add (string value, string? groupName = null, string? groupComment = null, string? symbolSuffix = null,
+			LlvmIrStringEncoding encoding = LlvmIrStringEncoding.UTF8, StringComparison comparison = StringComparison.Ordinal)
 		{
 			if (value == null) {
 				throw new ArgumentNullException (nameof (value));
 			}
 
-			LlvmIrStringVariable? stringVar;
-			if (stringSymbolCache.TryGetValue (value, out stringVar) && stringVar != null) {
+			return Add (new StringHolder (value, encoding, comparison), groupName, groupComment, symbolSuffix);
+		}
+
+		LlvmIrStringVariable Add (StringHolder holder, string? groupName = null, string? groupComment = null, string? symbolSuffix = null)
+		{
+			if (stringSymbolCache.TryGetValue (holder, out LlvmIrStringVariable? stringVar) && stringVar != null) {
 				return stringVar;
 			}
 
 			LlvmIrStringGroup? group;
 			string groupPrefix;
 			if (String.IsNullOrEmpty (groupName) || String.Compare ("str", groupName, StringComparison.Ordinal) == 0) {
 				group = defaultGroup;
-				groupPrefix = ".str";
+				groupPrefix = $".{defaultGroupName}";
```

Comment:
I guess this is where the "default string group" comes in?

What does this do to the resulting LLVM-IR?

Reply:
It allows us to "namespace" strings, as described above - their names mustn't clash between modules.